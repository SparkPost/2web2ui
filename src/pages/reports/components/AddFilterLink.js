import React, { useState } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import qs from 'qs';

import { addFilters } from 'src/actions/reportOptions';
import { PageLink } from 'src/components/links';
import { stringifyTypeaheadfilter } from 'src/helpers/string';
import {
  selectReportSearchOptions,
  selectSummaryChartSearchOptions,
} from 'src/selectors/reportSearchOptions';

export const AddFilterLink = ({
  //From parent
  newFilter,
  reportType,
  content,
  //From redux
  currentSearchOptions,
  addFilters,
}) => {
  const [isNewTabKeyPressed, setNewTabKeyPressed] = useState(false);

  const currentFilters = currentSearchOptions.filters || [];

  /**
   * Needs to first check if cmd/ctrl key is pressed when mouse is released
   * Then it's saved for use in the onClick handler. We need both because
   * onClick does not have the correct values for metaKey and ctrlKey
   **/
  const handleMouseUp = e => {
    setNewTabKeyPressed(e.metaKey || e.ctrlKey);
  };

  const handleClick = () => {
    if (!isNewTabKeyPressed) {
      addFilters([newFilter]);
    }
  };

  const mergedFilters = _.uniqWith(
    [...currentFilters, stringifyTypeaheadfilter(newFilter)],
    _.isEqual,
  );
  const newSearchOptions = { ...currentSearchOptions, filters: mergedFilters };
  const linkParams = qs.stringify(newSearchOptions, { arrayFormat: 'repeat' });
  const fullLink = `/reports/${reportType}/?${linkParams}`;

  return (
    <PageLink onMouseUp={handleMouseUp} onClick={handleClick} to={fullLink} replace>
      {content}
    </PageLink>
  );
};

const mapStateToProps = (state, props) => ({
  currentSearchOptions:
    props.reportType === 'summary'
      ? selectSummaryChartSearchOptions(state)
      : selectReportSearchOptions(state),
});

export default connect(mapStateToProps, { addFilters })(AddFilterLink);
