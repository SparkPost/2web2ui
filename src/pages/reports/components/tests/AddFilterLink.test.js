import React from 'react';
import { AddFilterLink } from '../AddFilterLink';
import { shallow } from 'enzyme';

describe('Add Filter Link', () => {
  const filters = ['Campaign: shiny new filter'];
  const baseProps = {
    newFilter: { id: 0, type: 'Subaccount', value: 'Primary Account (ID 0)' },
    reportType: 'summary',
    content: 'primary account',
    addFilters: jest.fn(),
    currentSearchOptions: {
      filters,
      metrics: ['count_something'],
    },
  };

  function subject(props) {
    return shallow(<AddFilterLink {...baseProps} {...props} />);
  }

  it('should render correctly for the summary report and include metrics in the link', () => {
    const wrapper = subject();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly for other reports and should not include metrics in the link', () => {
    const wrapper = subject({ reportType: 'bounce', currentSearchOptions: { filters } });
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle click correctly', () => {
    const wrapper = subject();
    wrapper.find('PageLink').simulate('mouseUp', {});
    wrapper.find('PageLink').simulate('click');

    expect(baseProps.addFilters).toHaveBeenCalledWith([
      { id: 0, type: 'Subaccount', value: 'Primary Account (ID 0)' },
    ]);
  });

  it('should handle click while holding down meta(cmd) key correctly', () => {
    const wrapper = subject();
    wrapper.find('PageLink').simulate('mouseUp', { metaKey: true });
    wrapper.find('PageLink').simulate('click');
    expect(baseProps.addFilters).not.toHaveBeenCalled();
  });

  it('should handle click while holding down control key correctly', () => {
    const wrapper = subject();
    wrapper.find('PageLink').simulate('mouseUp', { ctrlKey: true });
    wrapper.find('PageLink').simulate('click');
    expect(baseProps.addFilters).not.toHaveBeenCalled();
  });
});
