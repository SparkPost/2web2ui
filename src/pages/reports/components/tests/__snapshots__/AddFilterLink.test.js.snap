// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Filter Link should render correctly for other reports and should not include metrics in the link 1`] = `
<PageLink
  onClick={[Function]}
  onMouseUp={[Function]}
  replace={true}
  to="/reports/bounce/?filters=Campaign%3A%20shiny%20new%20filter&filters=Subaccount%3APrimary%20Account%20%28ID%200%29%3A0"
>
  primary account
</PageLink>
`;

exports[`Add Filter Link should render correctly for the summary report and include metrics in the link 1`] = `
<PageLink
  onClick={[Function]}
  onMouseUp={[Function]}
  replace={true}
  to="/reports/summary/?filters=Campaign%3A%20shiny%20new%20filter&filters=Subaccount%3APrimary%20Account%20%28ID%200%29%3A0&metrics=count_something"
>
  primary account
</PageLink>
`;
